tests/unit_tests/cpu/covariance_transport.cpp

```diff
@@ -127,50 +237,194 @@ GTEST_TEST(detray_propagation, helix_covariance_transport_cartesian2D) {
-            free_to_bound_jacobi =
-                c2.free_to_bound_jacobian(trfs[next_index], free_trk.vector());
+        // Free-to-bound jacobian at the destination surface
+        const free_to_bound_matrix free_to_bound_jacobi =
+            destination_frame.free_to_bound_jacobian(trf_1,
+                                                     free_trk_1.vector());
+
+        // Bound vector at the destination surface
+        const bound_vector bound_vec_1 =
+            destination_frame.free_to_bound_vector(trf_1, free_trk_1.vector());
 
         // Full jacobian
-        const typename cartesian2<transform3>::bound_matrix full_jacobi =
-            free_to_bound_jacobi * correction_term * transport_jacobi *
-            bound_to_free_jacobi;
+        const bound_matrix full_jacobi = free_to_bound_jacobi *
+                                         correction_term * transport_jacobi *
+                                         bound_to_free_jacobi;
+
+        // Update the covariance at the destination surface
+        const bound_matrix bound_cov_1 =
+            full_jacobi * bound_cov_0 *
+            matrix_operator().transpose(full_jacobi);
+
+        bound_track_parameters<transform3> ret;
+        ret.set_vector(bound_vec_1);
+        ret.set_covariance(bound_cov_1);
+
+        return ret;
+    }
 
-        // Update the covariance at the next surface
-        bound_cov =
-            full_jacobi * bound_cov * matrix_operator().transpose(full_jacobi);
+    private:
+    template <
+        typename I, typename M,
+        typename std::enable_if_t<
+            std::is_same_v<I, detail::helix_plane_intersector<intersection_t>>,
+            bool> = true>
+    intersection_t get_intersection(const detail::helix<transform3>& hlx,
+                                    const transform3_type& trf, const M& mask) {
+        return I{}(hlx, surface<>{}, mask, trf, mask_tolerance);
+    }
+
+    template <
+        typename I, typename M,
+        typename std::enable_if_t<
+            std::is_same_v<I, detail::helix_line_intersector<intersection_t>>,
+            bool> = true>
+    intersection_t get_intersection(const detail::helix<transform3>& hlx,
+                                    const transform3_type& trf, const M& mask) {
+        return I{}(hlx, surface<>{}, mask, trf, mask_tolerance);
+    }
+
+    template <typename I, typename M,
+              typename std::enable_if_t<
+                  std::is_same_v<
+                      I, detail::helix_cylinder_intersector<intersection_t>>,
+                  bool> = true>
+    intersection_t get_intersection(const detail::helix<transform3>& hlx,
+                                    const transform3_type& trf, const M& mask) {
+        return I{}(hlx, surface<>{}, mask, trf, mask_tolerance)[0];
+    }
+};
+
+using TestTypes = ::testing::Types<
+    std::pair<rectangle_type, detail::helix_plane_intersector<intersection_t>>,
```

Comment:
I doubt the ```propagate``` function needs to be templated on the intersector type? ```helix_intersector<intersection_t, mask_type>``` will give you the correct helix intersector.

So, this should work:
```
const intersection_t get_first_intersection(const std::array<intersection_t, 2> &inters) {
     return inters[0];
}
const intersection_t get_first_intersection(const intersection_t &inters) {
     return inters;
}

const intersection_t inters = get_first_intersection(helix_intersector<intersection_t, mask_t>{}(helix, surface, mask, trf));

```

Reply:
I am trying to do this and now I am trapped :/ Are you sure this is working code?

Reply:
Maybe i have to call impl type?

Reply:
Hm, I haven't tried to compile/run the lines above tbh. But smthing similar is running in the ```helix_intersection_kernel``` (I hope). What is the error/problem?

Reply:
nvm it is working after small modification